### PR TITLE
Fix KMeans++ initialisation slowness

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference issue

Relates to: #11877

#### What does this implement/fix?

The squared distances were computed in Python in a suboptimal way with a
complexity of O(n²). It took the overall running time of the function.

They are now computed using  `cdist`, making the overall algorithm fast.

### CODE REVIEW

**Comment by jjerphan** [comment]
Using the snippet provided by @Gabriel-p in #11877:

```python
import numpy as np
from sklearn.cluster import KMeans
from scipy.cluster.vq import kmeans2
import time as t

N = 1000
data = np.random.uniform(0., 1., (N, 4))

n_clusters = 50

s = t.time()
centroid, labels1 = kmeans2(
    data, n_clusters, minit='++', iter=10, thresh=1.e-4)
print("scipy:", t.time() - s)

s = t.time()
# KMeans
model = KMeans(n_clusters=n_clusters, init='k-means++', n_init=10, tol=1.e-4,
    max_iter=1000, verbose=0)
model.fit(data)
labels2 = model.labels_
print("sklearn", t.time() - s)
```

We previously got the following running times:
```
scipy: 3.926667865371704
sklearn 0.741915225982666
```

With this fix, we now get:
```
scipy: 0.016803741455078125
sklearn 0.741915225982666
```

**Comment by Gabriel-p** [comment]
Awesome work @jjerphan!

File: scipy/cluster/vq.py [hunk 1]
```diff
@@ -557,9 +557,7 @@ def _kpp(data, k):
             init[i, :] = data[np.random.randint(dims)]
 
         else:
-            D2 = np.array([min(
-                            [np.inner(init[j]-x, init[j]-x) for j in range(i)]
-                            ) for x in data])
+            D2 = (cdist(init[:i,:], data)**2).min(axis=0)
```
Comment (ewmoore) [inline scipy/cluster/vq.py:560]: Can you use `metric='sqeuclidean'` instead of squaring after the fact?

Comment (jjerphan) [reply scipy/cluster/vq.py]: Yes, thanks for pointing it out!
I just have implemented it; it works and has similar running times.

**Comment by jjerphan** [comment]
Shall I add some tests in for this issue?

**Comment by rgommers** [comment]
Thanks @jjerphan. This code already is covered by tests (I assume, would be good to check). Extending the benchmark in `benchmarks/benchmarks/cluster.py` to include the `init` keywords would be nice to make sure we detect future performance regressions.

**Comment by jjerphan** [comment]
>  This code already is covered by tests (I assume, would be good to check).

Yes, there is one test for the initialisation:
https://github.com/scipy/scipy/blob/787efe4d2829a1f9c177aa368e9d73202ee044e4/scipy/cluster/tests/test_vq.py#L245-L259

and several to test reproducible results.

**Comment by jjerphan** [comment]
@rgommers: I've separated the benchmark in two classes as only `kmeans2` support different initialisation with `init`. Let me know if this makes sense or if I should change it.

**Comment by rgommers** [comment]
LGTM, merged. Thanks @jjerphan!

**Comment by mdhaber** [comment]
@jjerphan I just added to the [release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.5.0/_edit) "Initialization of `scipy.cluster.vq.kmeans2` using `minit="++"` is now much faster". Please edit as you see fit!

**Comment by jjerphan** [comment]
Thank you @mdhaber : I've just added a few more words to what you've written.
